auth_ldap_attribute_sync/tests/test_auth_ldap_attribute_sync.py

```diff
@@ -0,0 +1,96 @@
+# Copyright (C) 2018 Brainbean Apps
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl.html).
+
+import logging
+from openerp.tests.common import TransactionCase
+from odoo import api, registry, SUPERUSER_ID
+
+_logger = logging.getLogger(__name__)
+
+
+class FakeLdapConnection(object):
+    def __init__(self, entries):
+        self.entries = entries
+
+    def simple_bind_s(self, dn, passwd):
+        pass
+
+    def search_st(self, dn, scope, ldap_filter, attributes, timeout):
+        if dn in self.entries:
+            return [(dn, dict(self.entries[dn]))]
+        return None
+
+    def unbind(self):
+        pass
+
+    def unbind_s(self):
+        pass
+
+    def __getattr__(self, name):
+        def wrapper():
+            raise Exception("'%s' is not mocked" % name)
+        return wrapper
+
+
+class TestAuthLdapAttributeSync(TransactionCase):
+    def test_auth_ldap_attribute_sync(self):
+        with registry(self.env.cr.dbname).cursor() as cr:
+            env = api.Environment(cr, SUPERUSER_ID, {})
+            company = env['res.company']._company_default_get()
+            env['res.company.ldap'].sudo().create({
+                'company': company.id,
+                'ldap_base': 'dc=example,dc=com',
+                'ldap_filter': '(uid=%s)',
+                'ldap_binddn': 'cn=bind,dc=example,dc=com',
+                'create_user': True,
+                'name_attribute': 'displayName',
+                'name_attribute_mode': 'initial',
+                'email_attribute': 'mail',
+                'email_attribute_mode': 'always',
+            })
+        
+        self.env['res.company.ldap']._patch_method(
```

Comment:
better use [mock](https://docs.python.org/3/library/unittest.mock.html) here

Reply:
```
        CompanyLDAP = self.env['res.company.ldap']
        User = self.env['res.users']

        user_id = False
        with patch.object(
            CompanyLDAP,
            '_connect',
            lambda self, conf: FakeLdapConnection({
                'dc=example,dc=com': {
                    'cn': ['User Name'],
                    'displayName': ['User Name'],
                    'mail': ['user@example.com'],
                }
            })
        ):
            user_id = User.authenticate(
                self.env.cr.dbname,
                'username',
                'password',
                {}
            )
        user = User.sudo().browse(user_id)
```
does not quite work for me, is that what you meant?

Reply:
check out this test, there they make a lot of use of the mock package: https://github.com/OCA/social/blob/11.0/mail_tracking_mailgun/tests/test_mailgun.py#L340